ng-dev/release/publish/actions/shared/branch-off-next-branch.ts

```diff
@@ -8,35 +8,46 @@
-import {ReleaseNotes, workspaceRelativeChangelogPath} from '../../notes/release-notes.js';
-import {
-  computeNewPrereleaseVersionForNext,
-  getReleaseNotesCompareVersionForNext,
-} from '../../versioning/next-prerelease-version.js';
-import {PullRequest, ReleaseAction} from '../actions.js';
+import {green, Log} from '../../../../utils/logging.js';
+import {workspaceRelativePackageJsonPath} from '../../../../utils/constants.js';
+import {ReleaseNotes, workspaceRelativeChangelogPath} from '../../../notes/release-notes.js';
+import {PullRequest} from '../../actions.js';
 import {
   getCommitMessageForExceptionalNextVersionBump,
   getReleaseNoteCherryPickCommitMessage,
-} from '../commit-message.js';
+} from '../../commit-message.js';
+import {CutNpmNextPrereleaseAction} from '../cut-npm-next-prerelease.js';
+import {CutNpmNextReleaseCandidateAction} from '../cut-npm-next-release-candidate.js';
+import {ActiveReleaseTrains} from '../../../versioning/active-release-trains.js';
 
 /**
- * Base action that can be used to move the next release-train into the feature-freeze or
- * release-candidate phase. This means that a new version branch is created from the next
- * branch, and a new pre-release (either RC or another `next`) is cut indicating the new phase.
+ * Base action that can be used to move the next release-train into the dedicated FF/RC
+ * release-train while also cutting a release to move the train into the `feature-freeze`
+ * or `release-candidate` phase.
+ *
+ * This means that a new version branch is created based on the next branch, and a new
+ * pre-release (either RC or another `next`) is cut- indicating the new phase.
  */
-export abstract class BranchOffNextBranchBaseAction extends ReleaseAction {
-  /**
-   * Phase which the release-train currently in the `next` phase will move into.
-   *
-   * Note that we only allow for a next version to branch into feature-freeze or
-   * directly into the release-candidate phase. A stable version cannot be released
-   * without release-candidate.
-   */
+export abstract class BranchOffNextBranchBaseAction extends CutNpmNextPrereleaseAction {
+  /** Phase which the release-train currently in the `next` phase will move into. */
   abstract newPhaseName: 'feature-freeze' | 'release-candidate';
 
+  // Instances of the action for cutting a NPM next pre-releases. We will re-use
+  // these for determining the "new versions" and "release notes comparison version".
+  // This helps avoiding duplication, especially since there are is some special logic.
+  private _nextPrerelease = new CutNpmNextPrereleaseAction(
+    new ActiveReleaseTrains({...this.active, releaseCandidate: null}),
+    this.git,
+    this.config,
+    this.projectDir,
+  );
+  private _rcPrerelease = new CutNpmNextReleaseCandidateAction(
+    new ActiveReleaseTrains({...this.active, releaseCandidate: this.active.next}),
+    this.git,
+    this.config,
+    this.projectDir,
+  );
```

Comment:
Note: I think this makes this action simpler to understand. Main reason it's done is to avoid relying on functions like  `computeNewPrelreleaseVersionForNext`. This method doesn't fit well into making the pre-release actions "re-usable". Conceptually it makes sense relying on these actions since it's basically what we perform too.